src/coord/src/coord.rs

```diff
@@ -3417,50 +3417,43 @@ impl Coordinator {
                     });
                 }
 
-                let mut candidate = if uses_ids.iter().any(|id| self.catalog.uses_tables(*id)) {
-                    // If the view depends on any tables, we enforce
-                    // linearizability by choosing the latest input time.
-                    self.get_local_read_ts()
-                } else {
-                    let upper = self.indexes.greatest_open_upper(index_ids.iter().copied());
-                    // We peek at the largest element not in advance of `upper`, which
-                    // involves a subtraction. If `upper` contains a zero timestamp there
-                    // is no "prior" answer, and we do not want to peek at it as it risks
-                    // hanging awaiting the response to data that may never arrive.
-                    //
-                    // The .get(0) here breaks the antichain abstraction by assuming this antichain
-                    // has 0 or 1 elements in it. It happens to work because we use a timestamp
-                    // type that meets that assumption, but would break if we used a more general
-                    // timestamp.
-                    if let Some(candidate) = upper.elements().get(0) {
-                        if *candidate > 0 {
-                            candidate.saturating_sub(1)
-                        } else {
-                            let unstarted = index_ids
-                                .into_iter()
-                                .filter(|id| {
-                                    self.indexes
-                                        .upper_of(id)
-                                        .expect("id not found")
-                                        .less_equal(&0)
-                                })
-                                .collect::<Vec<_>>();
-                            return Err(CoordError::IncompleteTimestamp(unstarted));
-                        }
+                // Initialize candidate to the minimum correct time.
+                let mut candidate = Timestamp::minimum();
+                candidate.advance_by(since.borrow());
+
+                // Compute a timestamp to which we should advance the candidate (if it is in
+                // advance).
+                let advance_to: Timestamp =
+                    if uses_ids.iter().any(|id| self.catalog.uses_tables(*id)) {
+                        // If the view depends on any tables, we enforce
+                        // linearizability by choosing the latest input time.
+                        self.get_local_read_ts()
```

Comment:
I'm not sure of the logic around here, but do we also need to make sure to advance the `read_ts` based on what we conclude here? Or is that logic elsewhere?

Reply:
We already weren't doing that. Previously we'd take `get_local_read_ts` and advance it to since, then not do anything if we did some advancing. I believe this PR comes up with the same timestamp just in a different order, but equally doesn't worry about telling table linearizability about it.

My take: table linearizability ensures that all table writes are seen by table reads, and that's it. It does not interact with other sources (which is the only way that we would bump the timestamp to be in advance of get_local_read_ts), and that can be done in our larger linearizability work.

Reply:
Ah, lemme give more detail then and lmk what you think:

1. `my_source` has a since of `10` and contains `source_val1` at `10`.
2. `my_table` has a since of `5` and contains `table_val1` at `5` and `table_val2` at `7`.

When we go and pick a timestamp of the union, we advance to `10` to get a legit read from `my_source`. When we query just `my_table` we start at `5` and that works great. Maybe .. are you maybe saying "we will always go at least to `7` because of `upper`"? That checks out, but seems like maybe a new constraint (must always advance to max `upper` to maintain linearizability).

Reply:
`my_table` cannot have a since of 5 if it has a value at 7.

Reply:
If the since for `my_table` is 5, the query timestamp will be at least `7` because of the call to `get_local_read_ts`. I actually think the table stuff and upper stuff should be separated, and could both trigger.

Reply:
(I deleted an incorrect comment.)

Reply:
I have a new take! In the case that the candidate is already in advance of the table read time, the peek will be put into pending until something closes that table timestamp, and once that read is served table writes can't be done before it. So I think we're ok?

Reply:
I've added a comment describing this. PTAL.